### ISSUE TITLE
RFC: ReadData abstract module

### ISSUE THREAD

**Issue by deanebarker** [statiqdev/Statiq#539]
I've put together an abstract module base called `ReadData` that can be extended to more easily create "document generation" modules -- those modules that initiate pipelines by turning data into `IDocument` objects.

For example:

    ReadObjectData(myCollectionOfPocoObjects);
    ReadSqlData("[connection string]", "SELECT * FROM Articles").WithContentProperty("Body");
    ReadXmlData("http://somesite.com/rss-feed", "//items")

It's an abstract module, so you would need to extend it, and override the `Execute` method to:

1. Get data from...somewhere
2. Set a protected field called `items` to an `IEnumerable<object>`
3. Return a call to `base.Execute`

The most basic example:

    public class ReadDataFromTheHeavens : ReadData<ReadDataFromTheHeavens>
    {
      public override IEnumerable<IDocument> Execute(IReadOnlyList<IDocument> inputs, IExecutionContext context)
      {
        items = TheLord.RainDataFromTheHeavens();
        // (It should really go without saying that The Lord is static.)

        return base.Execute(inputs, context);
      }
    }

The base method will turn whatever it finds in `items` into documents. With this, you can theoretically get data from anywhere -- if you can turn it into an object, this module will turn it into a document.

Using LINQ, you can create anonymous objects to set metadata up how you like:

    items = TheLord.RainDataFromTheHeavens().Select(x => new { SomeKey = x.SomeValue, SomeOtherKey = x.SomeOtherValue });

A fluent method allows you to specify which meta key should be the content of the document:

    ReadObjectData(myCollectionOfPocoObjects).WithContentKey("Body");

It's here:

https://gist.github.com/deanebarker/add3c2fee503e98b4dcffa36e2dcdd40

That contains examples, the base module itself, several extensions of it, and a "pattern sample" showing the basic method of extending it. The code is loquaciously commented.

I welcome all comments and feedback.

**Comment by daveaglick** [comment]
I'll take a closer look at this once I get the next release out, which is happening very soon. Related to #498.

**Comment by daveaglick** [comment]
Finally getting around to taking a look at this - here's some comments/questions:
* It's awesome
* No seriously, especially some of the concrete implementations like `ReadSqlData` - that's going to be huge
* Nice work on the TThis generic type parameter in the base so the fluent methods in the base all return the correct concrete type, that's one of my favorite tricks
* There's a lot of support for network/HTTP calls in the base abstract `ReadData` module - would it be appropriate to move all that networking stuff into a `ReadNetworkData` module?
* Along that same line of generalizing the base class, consider removing the `items` protected field and instead have a `protected abstract IEnumerable<object> GetItems()` in the base - then each specialized class (like `ReadSqlData`) can have all the stuff that's currently in the base for that specialization and the base `Execute()` would call `GetItems()` to get the data objects.

Looking forward to getting both the base module and all the implementations released - already thinking about a couple cool ideas for that SQL module.

**Comment by deanebarker** [comment]
FWIW, I am not a developer.

**Comment by deanebarker** [comment]
@daveaglick So, to clarify, you think we should do this:

    ReadData
      ReadNetworkData  <--HTTP util methods are here
        ReadWhateverData

Am I understanding that right?

**Comment by daveaglick** [comment]
Yep, exactly - take everything out of that abstract base class except the part that deals with `IEnumerable<object>` and turns it into a collection of output documents. Any logic dealing with how to get that collection of input objects can go in derived classes.

**Comment by daveaglick** [comment]
@deanebarker Starting to make a push for 1.0 and I'd like to get all of this in there - it's good stuff. I think I've got a pretty good idea where you were going and how to make it a bit cleaner. Mind if I run with what you've got in the gist, make a few changes, and then commit under your name?

**Comment by deanebarker** [comment]
@daveaglick Go for it. I was going to try to get something done next week, but if you have time, be my guest.

**Comment by daveaglick** [comment]
Had some fun with this today. Ended up with:
* `ReadDataModule` in `Wyam.Common` as the abstract base class for this functionality
* New `Xml`, `Sql`, `Objects`, and `AirTable` modules
* Did not bring over the network one since we already have `Download` which does essentially the same thing. However, I did port some of the features `Download` was missing from your network module.

I do have one question about the `AirTable` module - what should the query string look like with the API key. I couldn't find it in the gist and I'd like to bake it right into the module.

I think it worked out really, really well. Thanks for this! Take a look at the commit and let me know if you have any comments/questions.

**Comment by deanebarker** [comment]
You can pull the Airtable thing. That's pretty esoteric. It was only in there because I was using it. It's not well known.

**Comment by daveaglick** [comment]
One note about the removal of the network stuff from the base class and moving aspects of it into `Download`. For example, `Xml` doesn't allow you to specify an XML file or URL as it did in your original design.

I thought pretty hard about this and it struck me that that behavior was a bit counter to the whole modular pipeline pattern. Instead of the `Xml` module being able to pull data from specific location like the file system or the web, some other module should handle that and pass the data to the `Xml` module in the form of documents. So I made some changes to the base `ReadDataModule` class to allow derived classes to use input documents if it's appropriate. So now the `Xml` module can either have string-based XML content or can operate on the input documents.

In other words, your original example:

```
engine.Pipelines.Add("DocumentsFromXml",
        new ReadXmlData("https://gadgetopia.com/feed", "//item")
                .AddItemExtractionPath("url","./link")
                .AddQuerystringArgument("foo","bar")
                .ExcludeMetaKeys("link")
                .WithLimit(3)
                .WithContentProperty("description"),
        new ForEach(
                new Meta("WritePath", (doc, ctx) => { return string.Concat(new Uri(doc["url"].ToString()).Segments.Last().Trim("/".ToCharArray()), ".html"); })
        ),
        new WriteFiles()
);
```

Becomes (I renamed a couple of the methods too):

```
engine.Pipelines.Add("DocumentsFromXml",
	new Download(
		new DownloadRequest("https://gadgetopia.com/feed").WithQueryString("foo", "bar")
	),
        new Xml("//item")
                .WithMetadataPath("url","./link")
                .ExcludeKeys("link")
                .WithLimit(3)
                .WithContentKey("description"),
        new ForEach(
                new Meta("WritePath", (doc, ctx) => { return string.Concat(new Uri(doc["url"].ToString()).Segments.Last().Trim("/".ToCharArray()), ".html"); })
        ),
        new WriteFiles()
);
```

**Comment by deanebarker** [comment]
Curious about naming standards.

`ReadFiles` and `WriteFiles` begin with verbs and are actions -- _do these things_.  `Download` and `Xml` are nouns and are conceptual objects -- _these things are_.

I would rather see `DownloadData` and `ReadXml`, personally.

**Comment by daveaglick** [comment]
There's no real naming standard, and my thoughts on module naming have evolved a bit over time. Generally I like to keep them short - `Json`, `Yaml`, `Markdown`, etc. It's true that for most the verbs are implied and we're left with just a noun. I guess I think of it as *the Foo module* in which case it makes sense as a noun (though the form with a verb would certainly also work in that sense).

In the case of `ReadFiles` and `WriteFiles` the verbs are necessary to distinguish. If it weren't for the fact that we need two different modules to read and write to the file system, I'd probably have called the read module just `Files`.

If I were pressed to derive a standard I'd say it's:
- Modules are named with the action they perform followed by the type of data: `WriteFiles`, `ValidateMeta`
- If the module action is some variation of *process* it can be implied and just the type of data can be used: `Markdown`, `Yaml`
- If the action is specific enough that the type of data is implied, the result could be any type of data, or the action is used for control flow, just the action can be used: `Download`, `Branch`
